src/libopensc/pkcs15.c

```diff
@@ -475,20 +475,21 @@ parse_ddo(struct sc_pkcs15_card *p15card, const u8 * buf, size_t buflen)
 	LOG_TEST_RET(ctx, r, "DDO parsing failed");
 
 	if (asn1_ddo[1].flags & SC_ASN1_PRESENT) {
+		sc_file_free(p15card->file_odf);
 		p15card->file_odf = sc_file_new();
 		if (p15card->file_odf == NULL)
 			goto mem_err;
 		p15card->file_odf->path = odf_path;
 	}
 	if (asn1_ddo[2].flags & SC_ASN1_PRESENT) {
-		if (p15card->file_tokeninfo)
-			sc_file_free(p15card->file_tokeninfo);
+		sc_file_free(p15card->file_tokeninfo);
 		p15card->file_tokeninfo = sc_file_new();
 		if (p15card->file_tokeninfo == NULL)
 			goto mem_err;
 		p15card->file_tokeninfo->path = ti_path;
 	}
 	if (asn1_ddo[3].flags & SC_ASN1_PRESENT) {
+		sc_file_free(p15card->file_unusedspace);
```

Comment:
I think your fix is the correct one regarding the issue in question. however, I think we should still free the p15card's files since we are using the p15card object multiple times.

Reply:
Right.